### ISSUE TITLE
pipenv update on linux using python 3.10 leaves files in /tmp/ 

### ISSUE THREAD

**Issue by dbaxa** [pypa/pipenv#5180]
Be sure to check the existing issues (both open and closed!), and make sure you are running the latest version of Pipenv.

### Issue description

`pipenv update` leaves temp files in /tmp/ which should not persist/remain after python exits.

### Expected result

No temp files related to pipenv remain in /tmp/

### Actual result

pipenv temp files are left in `/tmp`

### Steps to replicate

`pipenv update`

-------------------------------------------------------------------------------

Please run `$ pipenv --support`, and paste the results here. Don't put backticks (`` ` ``) around it! The output already contains Markdown formatting.
pipenv --support

<details><summary>$ pipenv --support</summary>

Pipenv version: `'2022.7.4'`

Pipenv location: `'/home/user/.virtualenvs/myvenv/lib/python3.10/site-packages/pipenv'`

Python location: `'/home/user/.virtualenvs/myvenv/bin/python'`

Python installations found:

  - `3.10.4`: `/home/user/.virtualenvs/myvenv/bin/python`
  - `3.10.4`: `/home/user/.virtualenvs/myvenv/bin/python3`
  - `3.10.4`: `/home/user/.virtualenvs/myvenv/bin/python`
  - `3.10.4`: `/home/user/.virtualenvs/myvenv/bin/python3`
  - `3.10.4`: `/usr/bin/python3`
  - `3.10.4`: `/usr/bin/python`
  - `3.10.4`: `/bin/python3`
  - `3.10.4`: `/bin/python`
  - `2.7.18`: `/usr/bin/python2`
  - `2.7.18`: `/usr/bin/python2.7`
  - `2.7.18`: `/bin/python2`
  - `2.7.18`: `/bin/python2.7`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.10.4',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '5.15.0-41-generic',
 'platform_system': 'Linux',
 'platform_version': '#44-Ubuntu SMP Wed Jun 22 14:20:53 UTC 2022',
 'python_full_version': '3.10.4',
 'python_version': '3.10',
 'sys_platform': 'linux'}
```

System environment variables:

  - `SHELL`
  - `LANGUAGE`
  - `PWD`
  - `LOGNAME`
  - `XDG_SESSION_TYPE`
  - `MOTD_SHOWN`
  - `VIRTUALENVWRAPPER_SCRIPT`
  - `HOME`
  - `LANG`
  - `LS_COLORS`
  - `_VIRTUALENVWRAPPER_API`
  - `VIRTUAL_ENV`
  - `VIRTUALENVWRAPPER_WORKON_CD`
  - `SSH_CONNECTION`
  - `WORKON_HOME`
  - `LESSCLOSE`
  - `XDG_SESSION_CLASS`
  - `TERM`
  - `LESSOPEN`
  - `LIBVIRT_DEFAULT_URI`
  - `USER`
  - `VIRTUALENVWRAPPER_PROJECT_FILENAME`
  - `SHLVL`
  - `XDG_SESSION_ID`
  - `XDG_RUNTIME_DIR`
  - `PS1`
  - `SSH_CLIENT`
  - `MAVEN_OPTS`
  - `XDG_DATA_DIRS`
  - `BROWSER`
  - `PATH`
  - `VIRTUALENVWRAPPER_HOOK_DIR`
  - `DBUS_SESSION_BUS_ADDRESS`
  - `SSH_TTY`
  - `IRBRC`
  - `OLDPWD`
  - `GOPATH`
  - `ES_HOME`
  - `_`
  - `PIP_SHIMS_BASE_MODULE`
  - `PIP_DISABLE_PIP_VERSION_CHECK`
  - `PIP_PYTHON_PATH`
  - `PYTHONDONTWRITEBYTECODE`
  - `PYTHONFINDER_IGNORE_UNSUPPORTED`
  - `PIPENV_VERBOSITY`

Pipenv–specific environment variables:

Debug–specific environment variables:

  - `PATH`: `/home/user/.virtualenvs/myvenv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/user/.local/bin/`
  - `SHELL`: `/bin/bash`
  - `LANG`: `en_AU.UTF-8`
  - `PWD`: `/home/user/.virtualenvs/myvenv/lib/python3.10/site-packages/pipenv`
  - `VIRTUAL_ENV`: `/home/user/.virtualenvs/myvenv`

---------------------------

</details>

**Comment by dbaxa** [comment]
This issue seems to be related to the use of `vistir.path.create_tracked_tempfile` in `pipenv/environment.py`. It seems that inside `_TrackedTempfileWrapper.cleanup`  `self._finalizer.detach()` returns false and so the file is not deleted (see `pipenv/vendor/vistir/path.py`).

**Comment by dqkqd** [comment]
I've checked using `assert not os.path.isfile(tmpfile_path)` and saw that temporary file is properly deleted after the call `tmpfile.close()`. So `cleanup` function is working fine.
I think the issue is that `subprocess_run` create a new temporary file using the same filename without deleting it.
https://github.com/pypa/pipenv/blob/2d250f5b65a9fce24c77809a47ff7df48f79d61f/pipenv/environment.py#L387-L394

Should we call `os.remove` at the end of function to explicit delete that file? I could create a PR for this.

**Comment by matteius** [comment]
That would be great @dbaxa -- I suspect there are other places where sub process is not cleaning up tmp data as well beyond just this .json file.  Here is what my TMP folder looks like after clearing it out just the other day:
![image](https://user-images.githubusercontent.com/479892/181427599-6909463b-9ed1-430a-8e17-212ffe387a26.png)

**Comment by dqkqd** [comment]
Sorry for late response. I created a PR for this.

I don't have such folders as @matteius does. Unless I interrupt `pytest` while it's still running.
I use linux and have a lot of .pem files inside /tmp after running tests, but it should be fixed as described in https://github.com/python/cpython/issues/93353

So I think just make sure .json files are not leaking is enough.
